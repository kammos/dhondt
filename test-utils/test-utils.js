import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { calculatorReducer } from '../src/calculator';


const customRender = (ui, { initialState, store, ...options } = {}) => {
    const actualStore = store ?? configureStore({
        reducer: { calculator: calculatorReducer },
        preloadedState: initialState,
    });

    const Wrapper = ({ children }) => {
        return <Provider store={actualStore}>
            {children}
        </Provider>
    }

    return render(ui, { wrapper: Wrapper, ...options });
}


export * from '@testing-library/react';
export { customRender as render };