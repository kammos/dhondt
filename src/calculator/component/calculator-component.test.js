/**
 * @jest-environment jsdom
 */
import { render, waitFor } from 'test-utils';
import userEvent from '@testing-library/user-event'
import Calculator from '.';

describe("calculator-component", () => {
    it("should focus newly added party's name input", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const newPartyButton = ui.getByRole('button', {name: /new party/i});

        await user.click(newPartyButton);
        expect(ui.container).toMatchSnapshot();
        const newPartyNameInput = ui.getByLabelText(/party name/i);
        await waitFor(() => expect(newPartyNameInput).toHaveFocus());

        // do it again to check if it actually focuses the newest
        await user.click(newPartyButton);
        expect(ui.container).toMatchSnapshot();
        const newPartyNameInput2 = ui.getAllByLabelText(/party name/i)[1];
        await waitFor(() => expect(newPartyNameInput2).toHaveFocus());
    });

    it("should allow to add parties, each with unique and valid name", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const newPartyButton = ui.getByRole('button', {name: /new party/i});

        for (let i = 0; i < 10; i++) {
            await user.click(newPartyButton);
        }

        const partyNameInputs = ui.getAllByLabelText(/party name/i);
        const names = [];

        partyNameInputs.forEach(partyNameInput => {
            names.push(partyNameInput.value);
            expect(partyNameInput).toBeValid();
        });
    });

    it("should allow to set treshold field, and validate it", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const tresholdInput = ui.getByLabelText(/treshold/i);

        expect(tresholdInput).toBeValid();

        await user.dblClick(tresholdInput);
        await user.keyboard('{Delete}');

        expect(tresholdInput).toBeInvalid();
        expect(tresholdInput).toHaveErrorMessage("can not be empty");

        await user.keyboard('101');

        expect(tresholdInput).toBeInvalid();
        expect(tresholdInput).toHaveErrorMessage("at most 100");

        await user.keyboard('{Backspace}0');

        expect(tresholdInput).toBeValid();

        await user.keyboard('{Backspace}{Backspae}{Backspace}-10');

        expect(tresholdInput).toBeInvalid();
        expect(tresholdInput).toHaveErrorMessage("at least 0");
    });

    it("should allow to set treshold field, and validate it", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const seatsInput = ui.getByLabelText(/seats/i);
        await user.dblClick(seatsInput);
        await user.keyboard('{Delete}');

        expect(seatsInput).toBeInvalid();
        expect(seatsInput).toHaveErrorMessage("can not be empty");

        await user.keyboard('-10');

        expect(seatsInput).toBeInvalid();
        expect(seatsInput).toHaveErrorMessage("at least 0");

        await user.keyboard('{Backspace}{Backspace}{Backspace}10');

        expect(seatsInput).toBeValid();
    });

    it("should allow to set party name, and validate it", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const newPartyButton = ui.getByRole('button', {name: /new party/i});
        await user.click(newPartyButton);

        const partyNameInput = ui.getByLabelText(/party name/i);

        expect(partyNameInput).toBeValid();

        await user.dblClick(partyNameInput);
        await user.keyboard('{Delete}');

        expect(partyNameInput).toBeInvalid();
        expect(partyNameInput).toHaveErrorMessage("can not be empty");

        await user.keyboard('{Space}{Space}');

        expect(partyNameInput).toBeInvalid();
        expect(partyNameInput).toHaveErrorMessage("can not be empty");

        await user.keyboard('{Backspace}{Backspace}Partia');

        expect(partyNameInput).toBeValid();
    });

    it("should allow to set votes for party, and validate it", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const newPartyButton = ui.getByRole('button', {name: /new party/i});
        await user.click(newPartyButton);

        const votesInput = ui.getByLabelText(/votes/i);

        expect(votesInput).toBeValid();

        await user.dblClick(votesInput);
        await user.keyboard('{Delete}');

        expect(votesInput).toBeInvalid();
        expect(votesInput).toHaveErrorMessage("can not be empty");

        await user.keyboard('litery');

        expect(votesInput).toBeInvalid();
        expect(votesInput).toHaveErrorMessage("can not be empty");

        await user.dblClick(votesInput);
        await user.keyboard('{Delete}123');

        expect(votesInput).toBeValid();
    });

    it("should allow to remove parties", async () => {
        const user = userEvent.setup();
        const ui = render(<Calculator />);

        const newPartyButton = ui.getByRole('button', {name: /new party/i});
        await user.click(newPartyButton);

        const deleteButton = ui.getByRole('button', {name: /delete party/i});
        await user.click(deleteButton);

        expect(deleteButton).not.toBeInTheDocument();
    });
});