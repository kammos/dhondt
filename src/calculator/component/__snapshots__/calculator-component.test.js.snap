// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator-component should focus newly added party's name input 1`] = `
<div>
  <form
    autocomplete="off"
    class="form"
  >
    <div
      class="mainSettings"
    >
      <div
        aria-labelledby="main-settings-label"
        class="fieldset"
        role="group"
      >
        <div
          class="legend"
          id="main-settings-label"
        >
          Main settings
        </div>
        <div
          class="inputs"
        >
          <div
            class="container"
          >
            <label
              class="label number"
            >
              <div
                class="labelText"
              >
                Seats
              </div>
              <input
                aria-invalid="false"
                class="input"
                name="seats"
                type="number"
                value="100"
              />
            </label>
            <div
              aria-live="polite"
              class="error"
              hidden=""
              id="seats-error-0"
            />
          </div>
          <div
            class="container"
          >
            <label
              class="label number"
            >
              <div
                class="labelText"
              >
                Treshold
              </div>
              <input
                aria-invalid="false"
                class="input"
                name="treshold"
                type="number"
                value="0"
              />
            </label>
            <div
              aria-live="polite"
              class="error"
              hidden=""
              id="treshold-error-1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="parties"
    >
      <fieldset
        class="fieldset"
        style="border-color: hsl(0, 75%, 60%);"
      >
        <legend
          class="srOnly"
        >
          party 0
        </legend>
        <div
          class="container"
        >
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label undefined"
              >
                <div
                  class="labelText"
                >
                  Party name
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  data-partystart="0"
                  name="party name"
                  value="Red"
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="party name-error-2"
              />
            </div>
          </div>
          <div
            class="icons"
          >
            <button
              aria-label="delete party"
              class="deleteButton"
              type="button"
            >
              <div
                class="crossIcon"
              />
            </button>
          </div>
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label number"
              >
                <div
                  class="labelText"
                >
                  Votes
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  name="votes"
                  type="number"
                  value="100"
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="votes-error-3"
              />
            </div>
          </div>
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label checkbox"
              >
                <div
                  class="labelText"
                >
                  Ignores treshold
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  name="ignores-treshold"
                  type="checkbox"
                  value=""
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="ignores-treshold-error-4"
              />
            </div>
          </div>
        </div>
      </fieldset>
      <fieldset
        class="fieldset"
        style="border-color: hsl(110, 75%, 60%);"
      >
        <legend
          class="srOnly"
        >
          New party
        </legend>
        <button
          class="addParty"
          type="button"
        >
          New party
        </button>
      </fieldset>
    </div>
  </form>
</div>
`;

exports[`calculator-component should focus newly added party's name input 2`] = `
<div>
  <form
    autocomplete="off"
    class="form"
  >
    <div
      class="mainSettings"
    >
      <div
        aria-labelledby="main-settings-label"
        class="fieldset"
        role="group"
      >
        <div
          class="legend"
          id="main-settings-label"
        >
          Main settings
        </div>
        <div
          class="inputs"
        >
          <div
            class="container"
          >
            <label
              class="label number"
            >
              <div
                class="labelText"
              >
                Seats
              </div>
              <input
                aria-invalid="false"
                class="input"
                name="seats"
                type="number"
                value="100"
              />
            </label>
            <div
              aria-live="polite"
              class="error"
              hidden=""
              id="seats-error-0"
            />
          </div>
          <div
            class="container"
          >
            <label
              class="label number"
            >
              <div
                class="labelText"
              >
                Treshold
              </div>
              <input
                aria-invalid="false"
                class="input"
                name="treshold"
                type="number"
                value="0"
              />
            </label>
            <div
              aria-live="polite"
              class="error"
              hidden=""
              id="treshold-error-1"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="parties"
    >
      <fieldset
        class="fieldset"
        style="border-color: hsl(0, 75%, 60%);"
      >
        <legend
          class="srOnly"
        >
          party 0
        </legend>
        <div
          class="container"
        >
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label undefined"
              >
                <div
                  class="labelText"
                >
                  Party name
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  data-partystart="0"
                  name="party name"
                  value="Red"
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="party name-error-2"
              />
            </div>
          </div>
          <div
            class="icons"
          >
            <button
              aria-label="delete party"
              class="deleteButton"
              type="button"
            >
              <div
                class="crossIcon"
              />
            </button>
          </div>
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label number"
              >
                <div
                  class="labelText"
                >
                  Votes
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  name="votes"
                  type="number"
                  value="100"
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="votes-error-3"
              />
            </div>
          </div>
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label checkbox"
              >
                <div
                  class="labelText"
                >
                  Ignores treshold
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  name="ignores-treshold"
                  type="checkbox"
                  value=""
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="ignores-treshold-error-4"
              />
            </div>
          </div>
        </div>
      </fieldset>
      <fieldset
        class="fieldset"
        style="border-color: hsl(110, 75%, 60%);"
      >
        <legend
          class="srOnly"
        >
          party 1
        </legend>
        <div
          class="container"
        >
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label undefined"
              >
                <div
                  class="labelText"
                >
                  Party name
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  data-partystart="1"
                  name="party name"
                  value="Green"
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="party name-error-5"
              />
            </div>
          </div>
          <div
            class="icons"
          >
            <button
              aria-label="delete party"
              class="deleteButton"
              type="button"
            >
              <div
                class="crossIcon"
              />
            </button>
          </div>
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label number"
              >
                <div
                  class="labelText"
                >
                  Votes
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  name="votes"
                  type="number"
                  value="110"
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="votes-error-6"
              />
            </div>
          </div>
          <div
            class="input"
          >
            <div
              class="container"
            >
              <label
                class="label checkbox"
              >
                <div
                  class="labelText"
                >
                  Ignores treshold
                </div>
                <input
                  aria-invalid="false"
                  class="input"
                  name="ignores-treshold"
                  type="checkbox"
                  value=""
                />
              </label>
              <div
                aria-live="polite"
                class="error"
                hidden=""
                id="ignores-treshold-error-7"
              />
            </div>
          </div>
        </div>
      </fieldset>
      <fieldset
        class="fieldset"
        style="border-color: hsl(220, 75%, 60%);"
      >
        <legend
          class="srOnly"
        >
          New party
        </legend>
        <button
          class="addParty"
          type="button"
        >
          New party
        </button>
      </fieldset>
    </div>
  </form>
</div>
`;
